server/channels/store/storetest/channel_store.go

```diff
@@ -8065,3 +8067,144 @@ func testGetTeamForChannel(t *testing.T, ss store.Store) {
 	var nfErr *store.ErrNotFound
 	require.True(t, errors.As(err, &nfErr))
 }
+
+func testGetChannelsWithUnreadsAndWithMentions(t *testing.T, ss store.Store) {
+	setupMembership := func(
+		pushProp string,
+		withUnreads bool,
+		withMentions bool,
+		isDirect bool,
+		userId string,
+	) (model.Channel, model.ChannelMember) {
+		o1 := model.Channel{}
+		o1.TeamId = model.NewId()
+		o1.DisplayName = "Channel1"
+		o1.Name = NewTestId()
+		o1.Type = model.ChannelTypeOpen
+		if isDirect {
+			o1.Type = model.ChannelTypeDirect
+		}
+		o1.TotalMsgCount = 25
+		o1.LastPostAt = 12345
+		o1.LastRootPostAt = 12345
+		_, nErr := ss.Channel().Save(&o1, -1)
+		require.NoError(t, nErr)
+
+		m1 := model.ChannelMember{}
+		m1.ChannelId = o1.Id
+		m1.UserId = userId
+		m1.NotifyProps = model.GetDefaultChannelNotifyProps()
+		m1.NotifyProps[model.PushNotifyProp] = pushProp
+		_, err := ss.Channel().SaveMember(&m1)
+		if !withUnreads {
+			m1.MsgCount = o1.TotalMsgCount
+			m1.LastViewedAt = o1.LastPostAt
+		}
+		if withMentions {
+			m1.MentionCount = 5
+		}
+		require.NoError(t, err)
+
+		return o1, m1
+	}
+
+	type TestCase struct {
+		name           string
+		pushProp       string
+		userNotifyProp string
+		isDirect       bool
+		withUnreads    bool
+		withMentions   bool
+	}
+	ttcc := []TestCase{}
+
+	channelNotifyProps := []string{model.ChannelNotifyDefault, "other"}
+	userNotifyProps := []string{model.UserNotifyAll, model.UserNotifyMention, "other"}
+	boolRange := []bool{true, false}
+
+	nameTemplate := "pushProp: %s, userPushProp: %s, direct: %t, unreads: %t, mentions: %t"
+	for _, pushProp := range channelNotifyProps {
+		for _, userNotifyProp := range userNotifyProps {
+			for _, isDirect := range boolRange {
+				for _, withUnreads := range boolRange {
+					ttcc = append(ttcc, TestCase{
+						name:           fmt.Sprintf(nameTemplate, pushProp, userNotifyProp, isDirect, withUnreads, false),
+						pushProp:       pushProp,
+						userNotifyProp: userNotifyProp,
+						isDirect:       isDirect,
+						withUnreads:    withUnreads,
+						withMentions:   false,
+					})
+					if withUnreads {
+						ttcc = append(ttcc, TestCase{
+							name:           fmt.Sprintf(nameTemplate, pushProp, userNotifyProp, isDirect, withUnreads, true),
+							pushProp:       pushProp,
+							userNotifyProp: userNotifyProp,
+							isDirect:       isDirect,
+							withUnreads:    withUnreads,
+							withMentions:   true,
+						})
+					}
+				}
+			}
+		}
+	}
```

Comment:
@agnivade This may be a bit overkill (it adds right now 36 different tests) so let me know if you want to tackle this differently.

Also, doing the tests, the logic seems off (mainly due to the little reactivity to the channel notification props). I am almost sure the logic is the same as the logic we used to have, so this may require further inspection in a future ticket, but 0/5.

Reply:
My bad, I just discovered that even if the constant is different, the actual string is the same (`model.UserNotifyAll == model.ChannelNotifyAll`, `model.UserNotifyMention == model.ChannelNotifyMention`, `model.UserNotifyNone == model.ChannelNotifyNone`)

That being said, the number of tests goes up a bit (96 if my calculations are right). Let me know if this is desired or not, or what other approach would you take.

Reply:
I had to add other fixes in the test, so I added all cases. Once more, I understand that adding 96 tests to this feature may be overkill, so happy to consider other approaches.

Reply:
Yeah, since the strings are same, there isn't much use of so many cases. Mainly I wanted to test these 4 if conditions. If we can construct scenarios that execute them, that's good enough with me:

```go
if hasUnreads { // scenario 1
	channelsWithUnreads = append(channelsWithUnreads, channel.Id)
}

notify := channel.NotifyProps[model.PushNotifyProp]
if notify == model.ChannelNotifyDefault { // scenario 2
	notify = userNotifyProps[model.PushNotifyProp]
}
if notify == model.UserNotifyAll || channel.Type == string(model.ChannelTypeDirect) {
	if hasUnreads { // scenario 3
		channelsWithMentions = append(channelsWithMentions, channel.Id)
	}
} else if notify == model.UserNotifyMention {
	if hasMentions { // scenario 4
		channelsWithMentions = append(channelsWithMentions, channel.Id)
	}
}
```

Reply:
I can drastically reduce the number of tests by setting always userNotifyProps to `UserNotifyNone` unless the channel notify props are `ChannelNotifyDefault` That would leave things at 7(3 channel notify props + 4 user notify props)*2 (is direct)*2 (hasUnreads)*1.5 (hasMentions only if hasUnreads) = 42 tests.

I could go a bit further, like trying to separate scenario 2 from the others, so we don't have to test so many cases, or taking into consideration that `UserNotifyHere` and `UserNotifyNone` in this scenario are the same (reducing a bit the number of tests), but I feel going these directions we may end up testing the implementation and not the feature.

Another option is to just select "one particular case" for each scenario (around 10 tests more or less) and hope they are able to find any relevant issue in the future.

0/5 on the approach.